src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ThemeService.cs

```diff
@@ -215,145 +214,72 @@ private async Task<Either<ActionResult, Unit>> DeleteReleasesForPublications(IEn
                 publication.LatestPublishedReleaseVersionId = null;
             });
 
+            _contentDbContext.UpdateRange(publications);
             await _contentDbContext.SaveChangesAsync();
 
             // Some Content Db Releases may be soft-deleted and therefore not visible.
             // Ignore the query filter to make sure they are found
-            var releaseVersionIdsToDelete = await _contentDbContext
+            var releaseVersionsToDelete = await _contentDbContext
                 .ReleaseVersions
+                .AsNoTracking()
                 .IgnoreQueryFilters()
-                .Where(rv => publicationIds.Contains(rv.PublicationId))
-                .Select(rv => new IdAndPreviousVersionIdPair<string>(rv.Id.ToString(),
-                    rv.PreviousVersionId != null ? rv.PreviousVersionId.ToString() : null))
+                .Include(rv => rv.Release)
+                .Where(rv => rv.PublicationId == publicationId)
                 .ToListAsync();
 
-            var releaseVersionIdsInDeleteOrder = VersionedEntityDeletionOrderUtil
-                .Sort(releaseVersionIdsToDelete)
-                .Select(ids => Guid.Parse(ids.Id))
-                .ToList();
+            var releaseVersionsAndDataSetVersions = await releaseVersionsToDelete
+                .ToAsyncEnumerable()
+                .SelectAwait(async rv =>
+                {
+                    var releaseFileIds = await _contentDbContext
+                        .ReleaseFiles
+                        .Where(rf => rf.ReleaseVersionId == rv.Id)
+                        .Select(rf => rf.Id)
+                        .ToListAsync();
+
+                    var dataSetVersions = await _publicDataDbContext
+                        .DataSetVersions
+                        .Where(dsv => releaseFileIds.Contains(dsv.Release.ReleaseFileId))
+                        .ToListAsync();
+
+                    return new ReleaseVersionAndDataSetVersions(
+                        ReleaseVersion: rv,
+                        DataSetVersions: dataSetVersions);
+                })
+                .ToListAsync();
 
-            // Delete release entries in the Azure Storage ReleaseStatus table - if not it will attempt to publish
-            // deleted releases that were left scheduled
-            await _releasePublishingStatusRepository.RemovePublisherReleaseStatuses(releaseVersionIdsInDeleteOrder);
+            var releaseVersionIdsInDeleteOrder = releaseVersionsAndDataSetVersions
+                .Order(new DependentReleaseVersionDeleteOrderComparator())
```

Comment:
The main reason that this needed to be reworked like this is so that we ensure we delete DataSetVersions in the correct reverse order.  We can't delete a V2 before a V3 for instance because the V3 version has a set of changelog table entries that refer to entities that the V2 DataSetVersion owns and will attempt to remove via cascade delete when the DSV itself is deleted.

The original code that deleted in a certain order I think is actually potentially not that important, but I didn't see the harm of retaining the behaviour to delete in a "realistic" order.

Also regarding your point with preferring `Version` over `PreviousVersionId` to determine ReleaseVersion order, I 100% agree, and in fact you'll see that I have switched to using Version instead in the new ordering code.